body.tex

```diff
@@ -1,166 +1,466 @@
 
+The Butler datastore is located in an S3 Bucket and follows the same hierarchical structure that POSIX datastore does. Consumed and produced datasets are read and written directly from S3 as bytes, whenever possible, and only downloaded to temporary files for objects that are not serializable.
+Since the directory structure is preserved by the S3 datastore the entire data repository is trivially transferable between the cloud and a local filesystem.
 
+The Butler registry is a RDS PostgreSQL database that keeps track of all LSST science files.
+Files that are not managed by Data Butler are managed by the Condor File IO via Pegasus. These include the Data Butler configuration file, pipeline definition (Quantum) files, and the log files. The master instance serves as the staging site for these files. Various network protocols transfer the files between the master and the workers; instances do not share a filesystem.
 
+\subsection{Alternative architecture designs}
 
-\section{Some technicalities: author list and standard LSST references} 
+We discussed different architecture designs but did not pursue all of them due to time constraints.
 
-Thank you Tim Jenness and Wil O'Mullane for helping with templates! 
+One prominent idea was to use HTCondor utilities to transfer files with S3, rather than relying on Data Butler to communicate directly with S3. In this design HTCondor controls all file transfers, and all datasets are declared in the execution workflow. The proposed implementation was partially external to the LSST code base, and such a data transfer plug-in was added to the HTCondor code base recently. In each job there is a POSIX data repository for local data access, and a mechanism is needed to trigger and perform data ingestion at a central S3 data repository. This requires additional Generation 3 Middleware utilities for posix-S3 datastore copying and URL generation. These issues disfavored this architecture. A favorable benefit of this approach is the potential of bulk data transfer management optimizations.
 
-\subsection{The LSST LaTeX Classes}
+With respect to the database, we discussed the possibility of having local registries within individual jobs.
+Each job has a fast local SQLite registry.
+After each job finishes, the individual registry is then collated into the central database.
+Besides simplifying failure recovery, another benefit is to avoid potential bottlenecks related to large scale parallel SQL transactions by bundling multiple transactions together and handling database connections in a centralized manner.
+This envisions a fully shared-nothing architecture, but implementing this requires more thorough design work in conjunction with the Generation 3 Middleware development.
+Also, the current implementations of the Butler registry depend on conflict resolution syntax to resolve insertions of complex datasets, that may not work in this scenario.
```

Comment:
I don't think this is a problem; at worst it limits the performance, but fundamentally any insertions into the local SQLite can be done in the global registry as well.